In [4]:
import sqlite3 as sql

In [5]:
%%capture
%load_ext sql
%sql sqlite:///factbook.db

ModuleNotFoundError: No module named 'sql'

### Querying the table to access the data needed

In [2]:
%%sql
SELECT * FROM sqlite_master WHERE type='table'

UsageError: Cell magic `%%sql` not found.


In [3]:
%%sql
SELECT * FROM facts
LIMIT 5

UsageError: Cell magic `%%sql` not found.


In [26]:
%%sql
SELECT MIN(population),MAX(population),MIN(population_growth),MAX(population_growth) FROM facts

Done.


[(0, 1367485388, 0.0, 4.02)]

### Finding countries with the least population

In [27]:
%%sql
SELECT name FROM facts
GROUP BY name
HAVING MIN(population) = 0

Done.


[('Antarctica',)]

### Finding country with the highest population, this turns out to be the world population.

In [28]:
%%sql
SELECT name FROM facts
GROUP BY name
HAVING MAX(population) = 7256490011

Done.


[]

##### Antarctica has the least population, infact it has no population because it has no indigenous inhabitants, whlie the population of the world as when this project is being carried out is approximately 7.3 billion

In [7]:
%%sql
SELECT COUNT(*) FROM facts

Done.


[(260,)]

##### 260 countries were represented in this factbook, excluding the world population which makes it 261 rows in total.

### Calculating the average values for the population including the world population

In [8]:
%%sql
SELECT AVG(population),AVG(area) FROM facts

Done.


[(32242666.56846473, 555093.546184739)]

##### We observe that with the 'world' column, we won't be able to get the accurate result query for our countries, as the world row is more like a sum of all other countries attribute. And this affects the average populage and average area negatively.

### So the best thing to do here is to delete the world column.

In [9]:
%%sql
DELETE FROM facts
WHERE name = 'World'

0 rows affected.


[]

### Now, let's check the total row as against the former value(261)

In [10]:
%%sql
SELECT COUNT(*) FROM facts

Done.


[(260,)]

##### We see we have 1 row short from the original data

### Finding countries with population above average(62094928) and countries with area below the average(555093)

In [11]:
%%sql
SELECT name,area,population
FROM facts
WHERE population > (SELECT AVG(population) FROM facts) AND population > (SELECT AVG(area) FROM facts)

Done.


[('Afghanistan', 652230, 32564342),
 ('Algeria', 2381741, 39542166),
 ('Argentina', 2780400, 43431886),
 ('Bangladesh', 148460, 168957745),
 ('Brazil', 8515770, 204259812),
 ('Burma', 676578, 56320206),
 ('Canada', 9984670, 35099836),
 ('China', 9596960, 1367485388),
 ('Colombia', 1138910, 46736728),
 ('Congo, Democratic Republic of the', 2344858, 79375136),
 ('Egypt', 1001450, 88487396),
 ('Ethiopia', 1104300, 99465819),
 ('France', 643801, 66553766),
 ('Germany', 357022, 80854408),
 ('India', 3287263, 1251695584),
 ('Indonesia', 1904569, 255993674),
 ('Iran', 1648195, 81824270),
 ('Iraq', 438317, 37056169),
 ('Italy', 301340, 61855120),
 ('Japan', 377915, 126919659),
 ('Kenya', 580367, 45925301),
 ('Korea, South', 99720, 49115196),
 ('Mexico', 1964375, 121736809),
 ('Morocco', 446550, 33322699),
 ('Nigeria', 923768, 181562056),
 ('Pakistan', 796095, 199085847),
 ('Philippines', 300000, 100998376),
 ('Poland', 312685, 38562189),
 ('Russia', 17098242, 142423773),
 ('South Africa', 1219090, 53675563),
 ('Spain', 505370, 48146134),
 ('Sudan', 1861484, 36108853),
 ('Tanzania', 947300, 51045882),
 ('Thailand', 513120, 67976405),
 ('Turkey', 783562, 79414269),
 ('Uganda', 241038, 37101745),
 ('Ukraine', 603550, 44429471),
 ('United Kingdom', 243610, 64088222),
 ('United States', 9826675, 321368864),
 ('Vietnam', 331210, 94348835),
 ('European Union', 4324782, 513949445)]

### Finding countries with the highest ratios of water to land

In [12]:
%%sql
SELECT name,CAST(area_water as Float)/CAST(area_land as Float) ratio
FROM facts
ORDER BY ratio DESC
LIMIT 5

Done.


[('British Indian Ocean Territory', 905.6666666666666),
 ('Virgin Islands', 4.520231213872832),
 ('Puerto Rico', 0.5547914317925592),
 ('Bahamas, The', 0.3866133866133866),
 ('Guinea-Bissau', 0.2846728307254623)]

##### The first-five Countries with the highest ratios to land are British Indian Ocean Territory, Virgin Islands, Puerto Rico, Bahamas, The and  Guinea-Bissau

In [13]:
%%sql
SELECT name,CAST(area_water as Float)/CAST(area_land as Float) ratio
FROM facts
GROUP BY name
HAVING ratio >= 1

Done.


[('British Indian Ocean Territory', 905.6666666666666),
 ('Virgin Islands', 4.520231213872832)]

##### The countries with more water than land are two here: British Indian Ocean Territory (with ratio of  905.67) and Virgin Islands (with ratio of 4.52). This is so because countries with ratios greater than one imply having more land than water.

### Countries that will add the most people to their population the next year are those countries with migration rates higher than the average migration rate.

In [14]:
%%sql
SELECT name,AVG(migration_rate)
FROM facts

Done.


[('Southern Ocean', 3.433946188340806)]

In [15]:
%%sql
SELECT name,AVG(migration_rate)
FROM facts
WHERE migration_rate > (SELECT AVG(migration_rate) FROM facts)

Done.


[('Virgin Islands', 8.333698630136988)]

### Countries with higher death rate than birth rate. These are countries whose ratios of birth rate to death rate are greater than 1

In [16]:
%%sql
SELECT name,CAST(death_rate as Float)/CAST(birth_rate as Float) ratio
FROM facts
GROUP BY name
HAVING ratio >= 1
ORDER BY ratio DESC

Done.


[('Bulgaria', 1.6188340807174888),
 ('Serbia', 1.5044052863436124),
 ('Latvia', 1.431),
 ('Lithuania', 1.4128712871287128),
 ('Hungary', 1.3897379912663756),
 ('Monaco', 1.3894736842105262),
 ('Slovenia', 1.350356294536817),
 ('Ukraine', 1.3488805970149254),
 ('Germany', 1.3482880755608027),
 ('Saint Pierre and Miquelon', 1.3099730458221026),
 ('Romania', 1.3019693654266957),
 ('Croatia', 1.288888888888889),
 ('Greece', 1.2806004618937643),
 ('Belarus', 1.2485981308411216),
 ('Japan', 1.1992433795712485),
 ('Portugal', 1.1887810140237325),
 ('Russia', 1.1801724137931033),
 ('Estonia', 1.1798287345385348),
 ('Italy', 1.165903890160183),
 ('Bosnia and Herzegovina', 1.0992108229988726),
 ('Czech Republic', 1.0737279335410175),
 ('Moldova', 1.0491666666666666),
 ('Poland', 1.0462012320328542),
 ('Austria', 1.0010626992561105),
 ('European Union', 1.0)]